realm-transformer/src/main/groovy/io/realm/transformer/RealmOptionalAPITransformer.groovy

```diff
@@ -0,0 +1,183 @@
+        def classPool = createClassPool(inputs)
+
+        def pkgNameQuote = Pattern.quote("io.realm")
+        def pattern = Pattern.compile("^$pkgNameQuote\\.[^\\.]*\$")
+        logger.info "$pattern"
+        classNames.findAll {
+            if (it.matches(pattern)) {
+                logger.info "getPackageMethods $it"
+                return true
+            }
+            return false
+        }.each {
+            classPool.get(it).getDeclaredMethods().each {
+                def optionalAPIAnnotation = (OptionalAPI)it.getAnnotation(OptionalAPI.class)
+                if (optionalAPIAnnotation == null) {
+                    logger.info "$it.declaringClass.name $it.name doesn't have @OptionalAPI annotation"
+                } else if (optionalAPIAnnotation.dependencies().size() == 0) {
+                    throw new RuntimeException("$it.name doesn't have proper dependencies.")
+                } else if (!refClassNames.containsAll(optionalAPIAnnotation.dependencies())) {
+                    // Doesn't have enough dependencies, remove the API
+                    logger.info "RealmOptionalAPITransformer found $it.declaringClass.name $it.name with annotation. has been removed."
+                    it.declaringClass.removeMethod(it)
+                } else {
+                    logger.info "$it.declaringClass.name $it.name has all dependencies" + optionalAPIAnnotation.dependencies()
+                }
+            }
+        }
+
+        // Create outputs
+        classNames.each {
+            logger.info "  Modifying class ${it}"
+            def ctClass = classPool.getCtClass(it)
+            ctClass.writeFile(outputProvider.getContentLocation(
+                    'realm-optional-api', getInputTypes(), getScopes(), Format.DIRECTORY).canonicalPath)
+        }
+        logger.info "RealmOptionalAPITransformer end."
+    }
+
+    private Set<String> getClassNames(Collection<TransformInput> inputs) {
+        Set<String> classNames = new HashSet<String>()
+
+        logger.info "getClassNames starts"
+        inputs.each {
+            it.directoryInputs.each {
+                def dirPath = it.file.absolutePath
+                logger.info "dirPath $dirPath"
+                it.file.eachFileRecurse(FileType.FILES) {
+                    if (it.absolutePath.endsWith(SdkConstants.DOT_CLASS)) {
+                        def className =
+                                it.absolutePath.substring(
+                                        dirPath.length() + 1,
+                                        it.absolutePath.length() - SdkConstants.DOT_CLASS.length()
+                                ).replace(File.separatorChar, '.' as char)
+                        classNames.add(className)
+                    }
+                }
+            }
+
+            it.jarInputs.each {
+                def jarFile = new JarFile(it.file)
+                logger.info "jarFile " + jarFile.name
+                jarFile.entries().each {
```

Comment:
instead of `each` and `if` I would probably go with `filter` and `each` just to make the code flatter and have separation of concerns